### PR TITLE
Add runAsNonRoot securityContext check

### PR DESCRIPTION
**Pull request by sanderma**
This PR solves [Issue 49](https://github.com/open-policy-agent/gatekeeper-library/issues/49)

- It adds a check for the `runAsNonRoot` securityContext field.
- It adds tests to validate the rule is functioning.

### CODE REVIEW

**Comment by sanderma** [comment]
cla/linuxfoundation is signed.

**Comment by maxsmythe** [comment]
Thanks for the PR!

What is the expected config to invoke "run as non root"? The example constraint in the library currently looks like this:

```yaml
apiVersion: constraints.gatekeeper.sh/v1beta1
kind: K8sPSPAllowedUsers
metadata:
  name: psp-pods-allowed-user-ranges
spec:
  match:
    kinds:
      - apiGroups: [""]
        kinds: ["Pod"]
  parameters:
    runAsUser:
      rule: MustRunAs # MustRunAsNonRoot # RunAsAny 
      ranges:
        - min: 100
          max: 200
    runAsGroup:
      rule: MustRunAs # MayRunAs # RunAsAny 
      ranges:
        - min: 100
          max: 200
    supplementalGroups:
      rule: MustRunAs # MayRunAs # RunAsAny 
      ranges:
        - min: 100
          max: 200
    fsGroup:
      rule: MustRunAs # MayRunAs # RunAsAny 
      ranges:
        - min: 100
          max: 200
```

Did we want to have a "run as any non-root user" constraint look like the following?


```yaml
apiVersion: constraints.gatekeeper.sh/v1beta1
kind: K8sPSPAllowedUsers
metadata:
  name: psp-pods-allowed-user-ranges
spec:
  match:
    kinds:
      - apiGroups: [""]
        kinds: ["Pod"]
  parameters:
    runAsUser:
      rule:  MustRunAsNonRoot
```

File: library/pod-security-policy/users/template.yaml [hunk 1]
```diff
@@ -102,8 +102,17 @@ spec:
...
           not get_field_value("runAsUser", container, input.review)
+          not get_field_value("runAsNonRoot", container, input.review)
           params.rule != "RunAsAny"
-          msg := sprintf("Container %v is attempting to run without a required securityContext/runAsUser. Allowed runAsUser: %v", [container.name, params])
+          params.rule != "runAsNonRoot"
+          msg := sprintf("Container %v is attempting to run without a required securityContext/%v", [container.name, params.rule])
+        }
+
+        get_user_violation(params, container) = msg {
+          params.rule = "runAsNonRoot"
```
Comment (maxsmythe) [inline library/pod-security-policy/users/template.yaml:112]: I think this should be `MustRunAsNonRoot`?

File: library/pod-security-policy/users/template.yaml [hunk 2]
```diff
@@ -102,8 +102,17 @@ spec:
 
         get_user_violation(params, container) = msg {
           not get_field_value("runAsUser", container, input.review)
+          not get_field_value("runAsNonRoot", container, input.review)
           params.rule != "RunAsAny"
-          msg := sprintf("Container %v is attempting to run without a required securityContext/runAsUser. Allowed runAsUser: %v", [container.name, params])
+          params.rule != "runAsNonRoot"
```
Comment (maxsmythe) [inline library/pod-security-policy/users/template.yaml:107]: `MustRunAsNonRoot`?

File: library/pod-security-policy/users/template.yaml [hunk 3]
```diff
@@ -102,8 +102,17 @@ spec:
 
         get_user_violation(params, container) = msg {
           not get_field_value("runAsUser", container, input.review)
+          not get_field_value("runAsNonRoot", container, input.review)
```
Comment (maxsmythe) [inline library/pod-security-policy/users/template.yaml:105]: I don't think we need to test this here, as the `params.rule != "runAsNonRoot"` line below means that this check will only run for `MustRunAs`, which does require a specific user be provided.

`RunAsNonRoot` with an undefined user is checked in the extra `get_user_violation` you've added below.

File: src/pod-security-policy/users/src_test.rego [hunk 4]
```diff
@@ -213,7 +213,22 @@ test_user_input_container_run_as_rule_non_root_ignore_ranges {
...
-
+test_user_input_container_run_as_rule_non_root_runasnonroot_true {
+  input := {
+    "review": review(runAsNonRoot(true), [ctr("cont1", null)], null),
+    "parameters": runAsUser(rule("MustRunAsNonRoot", null))
+  }
+  results := violation with input as input
+  count(results) == 0
+}
+test_user_input_container_run_as_rule_non_root_runasnonroot_false {
```
Comment (maxsmythe) [inline src/pod-security-policy/users/src_test.rego:224]: I think you'd get an unexpected no violations if you added a test:

```
test_user_input_container_must_run_as_rule_runasnonroot_true {
  input := {
    "review": review(runAsNonRoot(true), [ctr("cont1", null)], null),
    "parameters": runAsUser(rule("MustRunAs", [range(100, 200)]))
  }
  results := violation with input as input
  count(results) == 1
}
```

In that scenario a user would expect a violation b/c they are requiring a specific user ID range.

**Review by maxsmythe — CHANGES_REQUESTED** [review]

File: src/pod-security-policy/users/src_test.rego (hunk 4 above)
Comment (sanderma) [reply src/pod-security-policy/users/src_test.rego]: Added test:
```rego
test_user_one_container_run_in_range_seccont_runasnonroot {
  input := { "review": review(null, [ctr("cont1", runAsNonRoot(true))], null), "parameters": user_mustrunas_100_200 }
  results := violation with input as input
  count(results) == 1
}
```

File: library/pod-security-policy/users/template.yaml (hunk 3 above)
Comment (sanderma) [reply library/pod-security-policy/users/template.yaml]: Like so?

If `runAsUser` is present, it should always be checked, the `accept_users` func sorts out if it's ok:

```rego
get_user_violation(params, container) = msg {
  rule := params.rule
  provided_user := get_field_value("runAsUser", container, input.review)
  not accept_users(rule, provided_user)
  msg := sprintf("Container %v is attempting to run as disallowed user %v. Allowed runAsUser: %v", [container.name, provided_user, params])
}
```

When we ask for a specific user range via `MustRunAs`, `runAsUser` field should be present:

```rego
get_user_violation(params, container) = msg {
  not get_field_value("runAsUser", container, input.review)
  params.rule = "MustRunAs"
  msg := sprintf("Container %v is attempting to run without a required securityContext/runAsUser", [container.name])
}
```

If `MustRunAsNonRoot` and user doesn't add `runAsUser`, or `runAsNonRoot`: fail:

```rego
get_user_violation(params, container) = msg {
  params.rule = "MustRunAsNonRoot"
  not get_field_value("runAsUser", container, input.review)
  not get_field_value("runAsNonRoot", container, input.review)
  msg := sprintf("Container %v is attempting to run without a required securityContext/runAsNonRoot or securityContext/runAsUser != 0", [container.name])
}
```

File: library/pod-security-policy/users/template.yaml (hunk 1 above)
Comment (sanderma) [reply library/pod-security-policy/users/template.yaml]: Fixed

File: library/pod-security-policy/users/template.yaml (hunk 2 above)
Comment (sanderma) [reply library/pod-security-policy/users/template.yaml]: Fixed

**Review by maxsmythe — APPROVED** [review]
This LGTM @EmandM, LGTY?

**Review by ritazh — APPROVED** [review]

File: library/pod-security-policy/users/template.yaml [hunk 5]
```diff
@@ -102,8 +102,15 @@ spec:
 
         get_user_violation(params, container) = msg {
           not get_field_value("runAsUser", container, input.review)
-          params.rule != "RunAsAny"
-          msg := sprintf("Container %v is attempting to run without a required securityContext/runAsUser. Allowed runAsUser: %v", [container.name, params])
+          params.rule = "MustRunAs"
+          msg := sprintf("Container %v is attempting to run without a required securityContext/runAsUser", [container.name])
```
Comment (sanderma) [inline library/pod-security-policy/users/template.yaml:106]: I see I left the param field out. The formatting was a bit off, but very useful.

**Comment by maxsmythe** [comment]
Ready to merge?

**Comment by sanderma** [comment]
> Ready to merge?

I don't see any other things to change at this point.

**Comment by maxsmythe** [comment]
Fair enough! Merging.

**Pull request merged**